src/engine.rs

```diff
@@ -251,37 +251,77 @@ impl Engine {
         origin: Address,
         value: U256,
         input: &[u8],
-    ) -> (ExitReason, Address) {
+    ) -> (ExitReason, SubmitResult) {
         let mut executor = self.make_executor();
         let address = executor.create_address(CreateScheme::Legacy { caller: origin });
         let (status, result) = (
             executor.transact_create(origin, value, Vec::from(input), u64::MAX),
             address,
         );
+        let used_gas = executor.used_gas();
         let (values, logs) = executor.into_state().deconstruct();
-        self.apply(values, logs, true);
-        (status, result)
+        self.apply(values, Vec::<Log>::new(), true);
+
+        let mut res_logs = Vec::new();
+        for log in logs {
+            res_logs.push(log.into());
+        }
+
+        let (res_status, used_gas) = match status {
+            ExitReason::Succeed(_) => (true, used_gas),
+            ExitReason::Revert(_) => (false, used_gas),
+            ExitReason::Error(_) | ExitReason::Fatal(_) => (false, 0),
+        };
+
+        let res = SubmitResult {
+            status: res_status,
+            gas_used: used_gas,
+            logs: res_logs,
+            result: result.0.to_vec(),
+        };
+        (status, res)
     }
 
-    pub fn call_with_args(&mut self, args: FunctionCallArgs) -> (ExitReason, Vec<u8>) {
+    pub fn submit_with_args(&mut self, args: FunctionCallArgs) -> (ExitReason, SubmitResult) {
         let origin = self.origin();
         let contract = Address(args.contract);
         let value = U256::zero();
-        self.call(origin, contract, value, args.input)
+        self.submit(origin, contract, value, args.input)
     }
 
-    pub fn call(
+    pub fn submit(
         &mut self,
         origin: Address,
         contract: Address,
         value: U256,
         input: Vec<u8>,
-    ) -> (ExitReason, Vec<u8>) {
+    ) -> (ExitReason, SubmitResult) {
         let mut executor = self.make_executor();
         let (status, result) = executor.transact_call(origin, contract, value, input, u64::MAX);
+        let used_gas = executor.used_gas();
         let (values, logs) = executor.into_state().deconstruct();
-        self.apply(values, logs, true);
-        (status, result)
+        // There is no way to return the logs to the NEAR log method as it only
+        // allows a return of UTF-8 strings.
+        self.apply(values, Vec::<Log>::new(), true);
+
+        let mut res_logs = Vec::new();
+        for log in logs {
+            res_logs.push(log.into());
+        }
+
+        let (res_status, used_gas) = match status {
+            ExitReason::Succeed(_) => (true, used_gas),
+            ExitReason::Revert(_) => (false, used_gas),
+            ExitReason::Error(_) | ExitReason::Fatal(_) => (false, 0),
+        };
+
+        let res = SubmitResult {
+            status: res_status,
+            gas_used: used_gas,
+            logs: res_logs,
+            result,
+        };
```

Comment:
Looks like a pretty big section of code duplication with `deploy_code`. Maybe it could be extracted out into a function and used in both places.

Reply:
Yeah, easily enough.

Reply:
@joshuajbouw Is this something you still plan to do?

Reply:
Not needed now with later changes that shortened that code drastically enough that it no longer is warranted.